### PR TITLE
Topic/fix pdef clear

### LINKED ISSUE #4767: Pdef quantization doesn't work when set as source to an Ndef
## Environment

* SuperCollider version: 3.10.3
* Operating system: Windows 10 Pro
* Other details (Qt version, audio driver, etc.):

## Steps to reproduce

```supercollider
// Get rid of old stuff
Ndef.clear; Pdef.removeAll; c = TempoClock(1.5);

// Create a "ping" metronome so we know where every 8th beat falls
Pdef(\ping, Pbind(\degree, Pseq([14, Pn(7,7)], inf), \dur, 1, \amp, 0.02)).play(c, quant: 8);

// Now create a quantized Ndef with a quantized Pdef as its source
p = Pbind(\degree, Pseq((0..3)), \dur, 0.125);
Ndef(\test).clock_(c).quant_(8).play;
Ndef(\test).source = Pdef(\pat, p).clock_(c).quant_(8);

// Play a minimal pattern that *should* start on a multiple-of-eight beat
Pdef(\pat).source = p.copy;
// (it doesn't, it seems to start right away, on the next beat...try it multiple times)

// In contrast, this works (not using Pdef but Pbind as Ndef source directly) -- the source changes at the quantization point:
Ndef(\test, p);
```

## Expected vs. actual behavior

If you're playing a `Pdef` directly, and when you set a Pdef as the source for an Ndef, the quant setting seems to be respected. However, if you set the Pdef source when it's not playing (but when it's set as source for a playing `Ndef`) the quant is **not** respected, and the new source pattern starts playing immediately on the next beat.

### PR DESCRIPTION
**Pull request by telephon**
## Purpose and Motivation

Where the old source pattern is shorter than the quant value, `Pdef` / `EventPatternProxy` would start the new pattern too early. This is because Pfindur only cuts and doesn’t extend.  ~~This change checks if the stream has ended and adds a silent event that holds the time until the new pattern comes in.~~ This change introduces a `filling` argument to `Pfindur`, which allows us to extend the stream until `dur` has been reached.

fixes #4767.

Also Pdef.clear didn't empty the dictionary as expected, this is fixed, too.

This supersedes #4072 which was an incomplete fix.

## Types of changes

Internal changes to embedInStream and constrainStream in EventPatternProxy

- Bug fix
- New Feature (as side effect)

New feature:
An extra argument to `Pfindur` was added. This was the least intrusive fix and it makes the solution available to other uses.

Possible inconsistency:
If someone would use a Pdef to stream a pattern that sometimes returns `nil` and then values again, this would change behaviour now (it would stop). But all other patterns behave like this, so it is an edge case. Opinions welcome. 

## To-do list

- [x] Code is tested
- [x] All tests are passing
- [x] This PR is ready for review

### CODE REVIEW

File: SCClassLibrary/Common/Streams/FilterPatterns.sc [hunk 1]
```diff
@@ -414,9 +414,9 @@ Pfinval : Pfin {
...
 
 Pfindur : FilterPattern {
-	var <>dur, <>tolerance;
-	*new { arg dur, pattern, tolerance = 0.001;
-		^super.new(pattern).dur_(dur).tolerance_(tolerance)
+	var <>dur, <>tolerance, <>filling;
+	*new { arg dur, pattern, tolerance = 0.001, filling;
+		^super.new(pattern).dur_(dur).tolerance_(tolerance).filling_(filling)
 	}
 	storeArgs { ^[dur,pattern,tolerance] }
```
Comment (totalgee) [inline SCClassLibrary/Common/Streams/FilterPatterns.sc]: Should `filling` be added here?

**Comment by totalgee** [comment]
Okay, so with my test cases, everything now works as expected. For my local installed SC version, I made a new `Pfindur2` (using the code from your version of `Pfindur` in this PR) and overloaded `EventPatternProxy.constrainStream` (having it call `Pfindur2` where you modified the call to `Pfindur`). I will keep this patch (these overloads) until a released version of SC contains these fixes.

Thanks very much -- I'll let you know if I run into any more related problems, but for now all looks great.

File: SCClassLibrary/Common/Streams/FilterPatterns.sc (hunk 1 above)
Comment (telephon) [reply SCClassLibrary/Common/Streams/FilterPatterns.sc]: well spotted!

**Comment by jamshark70** [comment]
> This change introduces a filling argument to Pfindur, which allows us to extend the stream until dur has been reached.

Could Psync be used for this? Psync help:

> If the total duration of the event pattern is shorter than the given maximum duration:
> 
> - Pfindur simply ends: no further time manipulation.
> - Psync inserts a rest to round the total duration up to the nearest multiple of the given quant.

If Psync will do the job, I'd suggest not to add a second way to do the same thing. (Sorry, I know you went to the trouble of unit testing the new feature. We've gone almost 2 decades without needing "filling" for Pfindur and it looks to me like there's a related class that will already fix this issue -- using the existing class would reduce future maintenance burden.)

**Comment by telephon** [comment]
>If Psync will do the job, I'd suggest not to add a second way to do the same thing. 

Oh dear, yes, and the worst is that I have written Psync myself. Let's see if that works.

**Comment by telephon** [comment]
Using Psync will be hacky, because it doesn't support fast forwarding. Psync relies on a quant value, while Pfindur has a definite duration value. 

So the new feature is similar, but different enough to justify the addition.

>We've gone almost 2 decades without needing "filling" for Pfindur 

It'll be useful in the future. And better than adding a new class as it seems to me.

**Comment by jamshark70** [comment]
> Psync relies on a quant value, while Pfindur has a definite duration value.

If you use Psync with maxdur == quant, there is no practical difference.

```
// Pfindur with filling
(
p = Pfindur(4, Pbind(
	\type, \rest,
	\dur, Pseries(1, 1, 6) * 0.1
), filling: Event.silent).trace.play;
)

( 'dur': 0.1, 'type': rest )
( 'dur': 0.2, 'type': rest )
( 'dur': 0.3, 'type': rest )
( 'dur': 0.4, 'type': rest )
( 'dur': 0.5, 'type': rest )
( 'dur': 0.6, 'type': rest )
( 'dur': Rest(1.0), 'delta': 1.9 )

// Psync with maxdur == quant
(
p = Psync(Pbind(
	\type, \rest,
	\dur, Pseries(1, 1, 6) * 0.1
), quant: 4, maxdur: 4).trace.play;
)

( 'dur': 0.1, 'type': rest )
( 'dur': 0.2, 'type': rest )
( 'dur': 0.3, 'type': rest )
( 'dur': 0.4, 'type': rest )
( 'dur': 0.5, 'type': rest )
( 'dur': 0.6, 'type': rest )
( 'dur': Rest(1.9), 'delta': 1.9 )

// Psync with maxdur == quant, pattern runs over the bar
// Psync *does* constrain to 4 beats (maxdur)
(
p = Psync(Pbind(
	\type, \rest,
	\dur, Pseries(1, 1, 20) * 0.1
), quant: 4, maxdur: 4).trace.play;
)

( 'dur': 0.1, 'type': rest )
( 'dur': 0.2, 'type': rest )
( 'dur': 0.3, 'type': rest )
( 'dur': 0.4, 'type': rest )
( 'dur': 0.5, 'type': rest )
( 'dur': 0.6, 'type': rest )
( 'dur': 0.7, 'type': rest )
( 'dur': 0.8, 'type': rest )
( 'dur': 0.9, 'delta': 0.4, 'type': rest )
```

> Using Psync will be hacky, because it doesn't support fast forwarding.

I don't find this to be true either. Psync simply accumulates durations. There is no reference to the clock's nextTimeOnGrid.

```
// Pfindur, filling, ffwd 2 beats
(
r = Prout {
	p = Pfindur(4, Pbind(
		\type, \rest,
		\dur, Pseries(1, 1, 6) * 0.1
	), filling: Event.silent).trace.asStream;

	p.fastForward(2, 0.001, Event.default).wait;

	p.embedInStream(Event.default);
}.play;
)

( 'dur': 0.1, 'type': rest )
( 'dur': 0.2, 'type': rest )
( 'dur': 0.3, 'type': rest )
( 'dur': 0.4, 'type': rest )
( 'dur': 0.5, 'type': rest )
( 'dur': 0.6, 'type': rest )
( 'dur': Rest(1.0), 'delta': 1.9 )

// Psync, maxdur == quant, ffwd
(
r = Prout {
	p = Psync(Pbind(
		\type, \rest,
		\dur, Pseries(1, 1, 6) * 0.1
	), quant: 4, maxdur: 4).trace.asStream;

	p.fastForward(2, 0.001, Event.default).wait;

	p.embedInStream(Event.default);
}.play;
)

( 'dur': 0.1, 'type': rest )
( 'dur': 0.2, 'type': rest )
( 'dur': 0.3, 'type': rest )
( 'dur': 0.4, 'type': rest )
( 'dur': 0.5, 'type': rest )
( 'dur': 0.6, 'type': rest )
( 'dur': Rest(1.9), 'delta': 1.9 )
```

I'm quite confident Psync already implements what is needed.

For that matter, Psync does one thing better than the proposed "filling" argument: `dur` and `delta` match with Psync, but "filling" retains the default from Event.silent.

**Comment by telephon** [comment]
Ah I see, this makes sense. The fact that for any `x < limit` `x.roundUp(limit) = limit` is enough of a guarantee.

**Comment by telephon** [comment]
Hm, somehow it doesn't solve the problem:

https://github.com/telephon/supercollider/tree/topic/fix-pdef-clear-using-Psync

I haven't found the reason yet.

**Comment by telephon** [comment]
I've made another PR that is an alternative to this one: #4792

**Comment by dyfer** [comment]
> I've made another PR that is an alternative to this one: #4792

@telephon should we close this then?

**Comment by totalgee** [comment]
@dyfer This one can be closed, as far as I'm concerned. The replacement #4792 has been merged into develop.
